examples/pytorch/language-modeling/run_fim.py

```diff
@@ -0,0 +1,831 @@
+        # Split by chunks of max_len.
+        result = {
+            k: [t[i : i + block_size] for i in range(0, total_length, block_size)]
+            for k, t in concatenated_examples.items()
+        }
+        result["labels"] = result["input_ids"].copy()
+        return result
+
+    # Get the FIM-specific token ids
+    prefix_tok_id = tokenizer.convert_tokens_to_ids(data_args.fim_prefix_token)
+    middle_tok_id = tokenizer.convert_tokens_to_ids(data_args.fim_middle_token)
+    suffix_tok_id = tokenizer.convert_tokens_to_ids(data_args.fim_suffix_token)
+    pad_tok_id = None
+
+    # If truncate_or_pad is on, also get pad token id
+    if data_args.truncate_or_pad:
+        pad_tok_id = tokenizer.convert_tokens_to_ids(data_args.pad_token)
+
+    # The two functions below perform the FIM transformation on the data (either PSM or SPM or PSM+SPM)
+    # Don't call fim_transform directly in .map()
+    # Adapted from https://github.com/loubnabnl/santacoder-finetuning/blob/main/fim.py#L22C13-L83
+    def fim_transform(example):
+        """
+        This function performs FIM transformation on a single example (list of tokens)
+        """
+        if np_rng.binomial(1, data_args.fim_rate):
+            boundaries = sorted(np_rng.randint(low=0, high=len(example) + 1, size=2))
+
+            prefix = example[: boundaries[0]]
+            middle = example[boundaries[0] : boundaries[1]]
+            suffix = example[boundaries[1] :]
+
+            if data_args.truncate_or_pad:
+                total_length = len(prefix) + len(middle) + len(suffix) + 3
+                diff = total_length - len(example)
+                if diff > 0:
+                    suffix = suffix[: max(0, len(suffix) - diff)]
+                elif diff < 0:
+                    suffix.extend([pad_tok_id] * (-diff))
+
+            if np_rng.binomial(1, data_args.fim_spm_rate):
+                # Apply Suffix-Prefix-Middle (SPM) transformation
+                transformed_example = [prefix_tok_id, suffix_tok_id] + suffix + [middle_tok_id] + prefix + middle
+            else:
+                # Apply Prefix-Suffix-Middle (PSM) transformation
+                transformed_example = [prefix_tok_id] + prefix + [suffix_tok_id] + suffix + [middle_tok_id] + middle
+        else:
+            transformed_example = example
+
+        return transformed_example
+
+    # Below function is the one you are supposed to call in the .map() function
+    def apply_fim(examples):
+        """
+        Apply FIM transformation to a batch of examples
+        """
+        fim_transform_ids = [fim_transform(ids) for ids in examples["input_ids"]]
+        examples["input_ids"] = fim_transform_ids
+        examples["labels"] = fim_transform_ids
+        # If your application requires custom attention mask, please adjust this function's below line
+        # since FIM transformation increases the number of tokens in input_ids and labels
```

Comment:
```suggestion
        # If your application requires custom attention mask, please adjust this function's below line.
        # Since FIM transformation increases the number of tokens in input_ids and labels
```